lib/src/command/add.dart

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:yaml_edit/yaml_edit.dart';
+
+import '../command.dart';
+import '../exceptions.dart';
+import '../exit_codes.dart' as exit_codes;
+import '../io.dart';
+import '../log.dart' as log;
+import '../package.dart';
+import '../package_info.dart';
+import '../package_name.dart';
+import '../pubspec.dart';
+import '../solver.dart';
+import '../utils.dart';
+
+/// Handles the `add` pub command.
+class AddCommand extends PubCommand {
+  @override
+  String get name => 'add';
+  @override
+  String get description => 'Add a dependency to the current package.';
+  @override
+  String get invocation => 'pub add <package> [options]';
+  @override
+  String get docUrl => 'https://dart.dev/tools/pub/cmd/pub-add';
+
+  bool get isDevelopment => argResults['dev'];
+  String get gitUrl => argResults['git-url'];
+  String get gitPath => argResults['git-path'];
+  String get gitRef => argResults['git-ref'];
+  String get hostUrl => argResults['hosted-url'];
+  String get path => argResults['path'];
+
+  AddCommand() {
+    argParser.addFlag('dev',
+        abbr: 'd',
+        negatable: false,
+        help: 'Adds packages to the development dependencies instead.');
+
+    argParser.addOption('git-url', help: 'Git URL of the package');
+    argParser.addOption('git-ref',
+        help: 'Git branch or commit to be retrieved');
+    argParser.addOption('git-path', help: 'Path of git package');
+    argParser.addOption('hosted-url', help: 'URL of package host server');
+    argParser.addOption('path', help: 'Local path');
+  }
+
+  @override
+  Future run() async {
+    if (argResults.rest.isEmpty) {
+      usageException('Must specify a package to be added.');
+    }
+
+    final hasGitOptions = gitUrl != null || gitRef != null || gitPath != null;
+    final hasHostOptions = hostUrl != null;
+
+    if ((path != null && hasGitOptions) ||
+        (hasGitOptions && hasHostOptions) ||
+        (hasHostOptions && path != null)) {
+      usageException('Packages must either be a git, hosted, or path package.');
+    }
+
+    final package = _parsePackage(argResults.rest.first);
```

Comment:
Why not simply have:
```dart
if (hasGitOptions) {
   ....
}
if (hasHostOptions) {
   ....
}
...
```

Or maybe put these utility methods in functions on this class, where they have access to the `gitUrl` parameters...